File: src/python/api/quiz.py
from flask import Blueprint, request, jsonify, Response
from datetime import datetime
from typing import List
import json
import pystache

from ..shared import Config
from ..engine.search import mongo_filter, parse_query, parse_timedelta
from ..engine.quiz import get_next_review, srs_map, repeat_review

api_quiz = Blueprint("quiz", __name__, url_prefix="/api/quiz")


@api_quiz.route("/treeview", methods=["POST"])
def r_quiz_treeview():
    def _recurse_parse_data(_data: List[dict], _deck: List[str], _depth: int = 0):
        do_loop = True

        while _depth < len(_deck) - 1:
            for c in _data:
                if c["name"] == _deck[_depth]:
                    c["children"] = c.get("children", [])
                    _recurse_parse_data(c["children"], _deck, _depth + 1)
                    do_loop = False
                    break

            _depth += 1

            if not do_loop:
                break

        if do_loop and _depth == len(_deck) - 1:
            full_name = "/".join(deck)
            deck_items = [d for d in all_items if d["deck"].startswith(full_name)]

            _data.append({
                "name": _deck[_depth],
                "fullName": full_name,
                "isOpen": _depth < 2,
                "stat": {
                    "new": len([d for d in deck_items if not d["nextReview"]]),
                    "leech": len([d for d in deck_items if d["srsLevel"] == 0]),
                    "due": len([d for d in deck_items if datetime.fromisoformat(d["nextReview"]) < now])
                }
            })

    db = Config.DB
    all_items = tuple(filter(mongo_filter(request.json["q"]), db.get_all()))
    now = datetime.now()

    deck_list = sorted(set(d["deck"] for d in all_items))
    deck_with_subdecks = []

    for d in deck_list:
        deck = d.split("/")
        for i, _ in enumerate(deck):
            subdeck = "/".join(deck[:i + 1])
            if subdeck not in deck_with_subdecks:
                deck_with_subdecks.append(subdeck)

    full_data = []
    for d in deck_with_subdecks:
        deck = d.split("/")
        _recurse_parse_data(full_data, deck)

    return jsonify(full_data)


@api_quiz.route("/", methods=["POST"])
def r_quiz_build():
    r = request.json
    and_cond = [parse_query(r["q"])]

    if r["deck"]:
        and_cond.append({"$or": [
            {"deck": {"$startswith": r["deck"] + "/"}},
            {"deck": r["deck"]}
        ]})

    if r["due"]:
        and_cond.append({"nextReview": {"$lte": str(datetime.now() + parse_timedelta(r["due"]))}})
    else:
        and_cond.append({"nextReview": {"$lte": str(datetime.now())}})

    db = Config.DB
    all_items = [c["id"] for c in filter(mongo_filter({"$and": and_cond}), db.get_all())]

    return jsonify({"ids": all_items})


@api_quiz.route("/render", methods=["POST"])
def r_quiz_render():
    db = Config.DB
    c = dict(db.conn.execute("""
    SELECT front, back, templateId, noteId
    FROM card
    WHERE id = ?
    """, (request.json["id"],)).fetchone())

    if c["front"].startswith("@md5\n"):
        t = db.conn.execute("""
        SELECT front, back
        FROM template
        WHERE id = ?
        """, (c["templateId"],)).fetchone()

        data = json.loads(db.conn.execute("""
        SELECT data
        FROM note
        WHERE id = ?
        """, (c["noteId"],)).fetchone()[0])

        c["front"] = pystache.render(t["front"], data)
        c["back"] = pystache.render(t["back"], data)

    return jsonify(c)


@api_quiz.route("/right", methods=["PUT"])
def r_quiz_right():
    db = Config.DB
    srs_level = db.conn.execute("""
    SELECT srsLevel FROM card WHERE id = ?
    """, (request.json["id"],)).fetchone()[0]

    if srs_level is None:
        srs_level = 0

    srs_level += 1

    if srs_level >= len(srs_map):
        srs_level = len(srs_map) - 1

    db.conn.execute("""
    UPDATE card
    SET
        srsLevel = ?,
        nextReview = ?
    WHERE id = ?
    """, (
        srs_level,
        str(get_next_review(srs_level))
    ))

    return Response(status=201)


@api_quiz.route("/wrong", methods=["PUT"])
def r_quiz_wrong():
    db = Config.DB
    srs_level = db.conn.execute("""
        SELECT srsLevel FROM card WHERE id = ?
        """, (request.json["id"],)).fetchone()[0]

    if srs_level is None:
        srs_level = 1

    srs_level -= 1

    if srs_level < 0:
        srs_level = 0

    db.conn.execute("""
        UPDATE card
        SET
            srsLevel = ?,
            nextReview = ?
        WHERE id = ?
        """, (
        srs_level,
        str(repeat_review())
    ))

    return Response(status=201)

File: src/node/api/quiz.ts
import { SearchParser } from "../engine/search";
import moment from "moment";
import { Router } from "express";
import g from "../config";
import { pp } from "../util";

interface ITreeViewStat {
    new: number;
    leech: number;
    due: number;
}

export interface ITreeViewItem {
    name: string;
    fullName: string;
    isOpen: boolean;
    children?: ITreeViewItem[];
    stat: ITreeViewStat;
}

const router = Router();

router.post("/", (req, res) => {
    const db = g.DB!;
    const parser = new SearchParser();
    const andCond = [];

    const parseResult = parser.doParse(req.body.q);
    if (parseResult) {
        andCond.push(parseResult.cond);
    }

    if (req.body.deck) {
        andCond.push({$or: [
            {deck: {$startswith: `${req.body.deck}/`}},
            {deck: req.body.deck}
        ]})
    }

    if (req.body.type !== "all") {
        const type: string = req.body.type;
        if (type === "due") {
            andCond.push({nextReview: {$lte: moment().toISOString()}});
        } else if (type === "leech") {
            andCond.push({
                srsLevel: 0
            })
        } else if (type === "new") {
            andCond.push({
                nextReview: {$exists: false}
            })
        } else {
            andCond.push({$or: [
                {nextReview: {$exists: false}},
                {nextReview: {$lte: moment().toISOString()}}
            ]});
        }
    }

    const due = req.body.due;
    if (due) {
        const m = /(-?\d+(?:\.\d+)?\S+)/.exec(due);
        if (m) {
            andCond.push({nextReview: {$lte: moment().add(parseFloat(m[1]), m[2] as any).toISOString()}})
        } else {
            andCond.push({$or: [
                {nextReview: {$exists: false}},
                {nextReview: {$lte: moment().toISOString()}}
            ]})
        }
    }

    return res.json({ids: db.parseCond({
        cond: {$and: andCond},
        fields: new Set(["deck"])
    }, {
        fields: ["id"]
    }).data.map((c) => c.id)});
});

router.post("/render", (req, res) => {
    const db = g.DB!;
    return res.json(db.render(req.body.id));
});

router.post("/treeview", (req, res) => {
    function recurseParseData(data: ITreeViewItem[], deck: string[], _depth = 0) {
        let doLoop = true;

        while (_depth < deck.length - 1) {
            for (const c of data) {
                if (c.name === deck[_depth]) {
                    c.children = c.children || [];
                    recurseParseData(c.children, deck, _depth + 1);
                    doLoop = false;
                    break;
                }
            }

            _depth++;

            if (!doLoop) {
                break;
            }
        }

        if (doLoop && _depth === deck.length - 1) {
            const fullName = deck.join("/");
            const thisDeckData = deckData.filter((d) => d.deck === fullName || d.deck.indexOf(`${fullName}/`) === 0);

            data.push({
                name: deck[_depth],
                fullName,
                isOpen: _depth < 2,
                stat: {
                    new: thisDeckData.filter((d) => !d.nextReview).length,
                    leech: thisDeckData.filter((d) => d.srsLevel === 0).length,
                    due: thisDeckData.filter((d) => d.nextReview && moment(d.nextReview).toDate() < now).length
                }
            });
        }
    }

    const db = g.DB!;

    const rSearch = new SearchParser();
    const cond = rSearch.doParse(req.body.q);

    const deckData = db.parseCond(cond || {}, {
        fields: ["nextReview", "srsLevel", "deck"]
    }).data;

    const now = new Date();

    const deckList: string[] = deckData.map((d: any) => d.deck);
    const deckWithSubDecks: string[] = [];

    deckList.filter((d, i) => deckList.indexOf(d) === i).sort().forEach((d) => {
        const deck = d.split("/");
        deck.forEach((seg, i) => {
            const subDeck = deck.slice(0, i + 1).join("/");
            if (deckWithSubDecks.indexOf(subDeck) === -1) {
                deckWithSubDecks.push(subDeck);
            }
        });
    });

    const fullData = [] as ITreeViewItem[];
    deckWithSubDecks.forEach((d) => {
        const deck = d.split("/");
        recurseParseData(fullData, deck);
    });

    return res.json(fullData);
});

router.put("/right", (req, res) => {
    const db = g.DB!;
    return res.json(db.markRight(req.params.id));
});

router.put("/wrong", (req, res) => {
    const db = g.DB!;
    return res.json(db.markWrong(req.params.id));
});

export default router;
